iothub/device/src/Transport/Amqp/AmqpTransportHandler.cs

```diff
@@ -522,19 +519,23 @@ private async Task DisposeMessageAsync(string lockToken, AmqpIoTDisposeActions o
 
         private void TwinMessageListener(Twin twin, string correlationId, TwinCollection twinCollection)
         {
-            if (correlationId != null)
+            if (correlationId == null)
             {
-                // It is a GET, just complete the task.
-                TaskCompletionSource<Twin> task;
-                if (_twinResponseCompletions.TryRemove(correlationId, out task))
-                {
-                    task.SetResult(twin);
-                }
+                // This is desired property updates, so call the callback with TwinCollection.
+                _onDesiredStatePatchListener(twinCollection);
             }
             else
             {
-                // It is a PATCH, just call the callback with the TwinCollection
-                _onDesiredStatePatchListener(twinCollection);
+                if(correlationId.StartsWith(AmqpTwinMessageType.Get.ToString(), StringComparison.OrdinalIgnoreCase) ||
+                    correlationId.StartsWith(AmqpTwinMessageType.Patch.ToString(), StringComparison.OrdinalIgnoreCase))
+                {
+                    // For Get and Patch, complete the task.
+                    TaskCompletionSource<Twin> task;
+                    if (_twinResponseCompletions.TryRemove(correlationId, out task))
```

Comment:
What should we do if we don't find that correlation Id in the dictionary? Would that warrant logging a warning?

Reply:
That's why it is a try remove. No logging.
That should never happen though. correlation Id is set by SDK on a request, this is where we are basically completing a task associated with this Id.

Reply:
Do you know how many bugs have occurred in history when the programmer said "that should never happen"?

You have to consider that not everything will happen as you expect. What does it hurt to throw a log statement in for this event, even if it may never happen? If it does, you'll be happy to have the logs.

Reply:
added logging